eras/mary/impl/src/Cardano/Ledger/Mary/TxBody.hs

```diff
@@ -256,6 +272,35 @@ instance Crypto c => EraTxBody (MaryEra c) where
     lensMaryTxBodyRaw atbrCerts $ \txBodyRaw certs -> txBodyRaw {atbrCerts = certs}
   {-# INLINEABLE certsTxBodyL #-}
 
+  upgradeTxBody atb = do
+    certs <- traverse upgradeTxCert (atbCerts atb)
+    pure $
+      MaryTxBody
+        { mtbInputs = atbInputs atb
+        , mtbOutputs = upgradeTxOut <$> atbOutputs atb
+        , mtbCerts = certs
+        , mtbWithdrawals = atbWithdrawals atb
+        , mtbTxFee = atbTxFee atb
+        , mtbValidityInterval = atbValidityInterval atb
+        , mtbUpdate = upgradeUpdate <$> atbUpdate atb
+        , mtbAuxDataHash = atbAuxDataHash atb
+        , mtbMint = mempty
+        }
+    where
+      upgradeUpdate :: Update (AllegraEra c) -> Update (MaryEra c)
+      upgradeUpdate (Update pp epoch) = Update (upgradeProposedPPUpdates pp) epoch
+
+      upgradeProposedPPUpdates ::
```

Comment:
I get a feeling that a more general version of such function could be placed into `Cardano.Ledger.Shelley.PParams` and reused in almost all places (except Babbage)